### PR TITLE
Document the return types of functions EDATE, EOMONTH, DATEVALUE, TIMEVALUE

### LINKED ISSUE #1101: Document the differences in return value type for functions EDATE, EOMONTH, DATEVALUE, TIMEVALUE
### Description

| Function | ODFF return type | Excel auto-formatting | Google Sheets auto-formatting | return type in HyperFormula |
|---|---|---|---|---|
| EDATE | Number | Number | Date | NUMBER_DATE |
| EOMONTH | Number | Number | Date | NUMBER_DATE |
| DATEVALUE | Date | Number | Number | NUMBER_DATE |
| TIMEVALUE | Time | Number | Number | NUMBER_TIME |

### Links

https://github.com/handsontable/hyperformula/issues/1083

### Type of issue

- [x] Missing documentation
- [ ] Error in the documentation
- [ ] Request for a demo
- [ ] Request for a tutorial, guide etc.
- [ ] Other issues, suggestions or ideas

### PR DESCRIPTION
**Pull request by sequba**
### Context

We decided to keep our return types of functions EDATE, EOMONTH, DATEVALUE, TIMEVALUE regardless of the inconsistencies)
- describe the differences between HF, Excel and G-Sheets in _Runtime differences with Microsoft Excel and Google Sheets_ guide
- describe the differences between HF and ODFF in _Built-in functions_
- Add unit tests that verify the return types

### How did you test your changes?

- [x] unit test pass
- [x] local docs build

### Types of changes

- [ ] Breaking change (a fix or a feature because of which an existing functionality doesn't work as expected anymore)
- [ ] New feature or improvement (a non-breaking change that adds functionality)
- [ ] Bug fix (a non-breaking change that fixes an issue)
- [ ] Additional language file, or a change to an existing language file (translations)
- [x] Change to the documentation

### Related issues:
Fixes #1101

### Checklist:

- [ ] My change is compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard.
- [ ] My change is compatible with Microsoft Excel.
- [ ] My change is compatible with Google Sheets.
- [x] My code follows the code style of this project.
- [ ] I described my changes in the [CHANGELOG.md](https://github.com/handsontable/hyperformula/blob/master/CHANGELOG.md) file.
- [ ] My changes require a documentation update.
- [ ] My changes require a migration guide.

### CODE REVIEW

File: docs/guide/built-in-functions.md [hunk 1]
```diff
@@ -68,34 +68,34 @@ Total number of functions: **{{ $page.functionsCount }}**
...
+| EDATE            | Shifts the given startdate by given number of months and returns it as the number of full days since [nullDate](../api/interfaces/configparams.md#nulldate).<br>Return type: `CellValueDetailedType.NUMBER_DATE` (non-compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard)                                                | EDATE(Startdate, Months)                            |
+| EOMONTH          | Returns the date of the last day of a month which falls months away from the start date. Returns the value in the form of number of full days since [nullDate](../api/interfaces/configparams.md#nulldate).<br>Return type: `CellValueDetailedType.NUMBER_DATE` (non-compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard) | EOMONTH(Startdate, Months)                          |
```
Comment (warpech) [inline docs/guide/built-in-functions.md:80]: I think the presented note is too horrifying.

WDYT about keeping the long explanation on the other page, and replacing the text here with a simple link:

> [See compatibility notes](list-of-differences.md)

Or even a footnote.

---

I think that, in fact, we could add this for *all* or *none* of the functions listed here that have an entry on the "List of differences" page.

File: docs/guide/list-of-differences.md [hunk 2]
```diff
@@ -17,6 +17,10 @@ See a full list of differences between HyperFormula, Microsoft Excel, and Google
...
 | SPLIT function                                     | =SPLIT("Lorem ipsum dolor", 0)                                            | This function works differently from Google Sheets version but should be sufficient to achieve the same functionality in most scenarios. Read SPLIT function description on [the Built-in Functions page](built-in-functions.md#text). | Different syntax and return value.                                                                                     | No such function.                                        |
+| DATEVALUE function                                 | =DATEVALUE("25/02/1991")                                                  | Type of the returned value: `CellValueDetailedType.NUMBER_DATE` (complient with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard)            | Cell auto-formatted as **regular number**                                                                              | Cell auto-formatted as **regular number**                |
```
Comment (warpech) [inline docs/guide/list-of-differences.md:20]:
```suggestion
| DATEVALUE function                                 | =DATEVALUE("25/02/1991")                                                  | Type of the returned value: `CellValueDetailedType.NUMBER_DATE` (compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard)            | Cell auto-formatted as **regular number**                                                                              | Cell auto-formatted as **regular number**                |
```

File: docs/guide/list-of-differences.md [hunk 3]
```diff
@@ -17,6 +17,10 @@ See a full list of differences between HyperFormula, Microsoft Excel, and Google
...
+| TIMEVALUE function                                 | =TIMEVALUE("14:31")                                                       | Type of the returned value: `CellValueDetailedType.NUMBER_TIME` (complient with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard)            | Cell auto-formatted as **regular number**                                                                              | Cell auto-formatted as **regular number**                |
```
Comment (warpech) [inline docs/guide/list-of-differences.md:21]:
```suggestion
| TIMEVALUE function                                 | =TIMEVALUE("14:31")                                                       | Type of the returned value: `CellValueDetailedType.NUMBER_TIME` (compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard)            | Cell auto-formatted as **regular number**                                                                              | Cell auto-formatted as **regular number**                |
```

File: docs/guide/list-of-differences.md [hunk 4]
```diff
@@ -17,6 +17,10 @@ See a full list of differences between HyperFormula, Microsoft Excel, and Google
...
+| EDATE function                                     | =EDATE(DATE(2019, 7, 31), 1)                                              | Type of the returned value: `CellValueDetailedType.NUMBER_DATE`                                                                                                                                                                        | Cell auto-formatted as **date**                                                                                        | Cell auto-formatted as **regular number**                |
```
Comment (warpech) [inline docs/guide/list-of-differences.md:22]:
```suggestion
| EDATE function                                     | =EDATE(DATE(2019, 7, 31), 1)                                              | Type of the returned value: `CellValueDetailedType.NUMBER_DATE` (non-compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard)            | Cell auto-formatted as **date**                                                                                        | Cell auto-formatted as **regular number**                |
```

File: docs/guide/list-of-differences.md [hunk 5]
```diff
@@ -17,6 +17,10 @@ See a full list of differences between HyperFormula, Microsoft Excel, and Google
...
+| EOMONTH function                                   | =EOMONTH(DATE(2019, 7, 31), 1)                                            | Type of the returned value: `CellValueDetailedType.NUMBER_DATE`                                                                                                                                                                        | Cell auto-formatted as **date**                                                                                        | Cell auto-formatted as **regular number**                |
```
Comment (warpech) [inline docs/guide/list-of-differences.md:23]:
```suggestion
| EOMONTH function                                   | =EOMONTH(DATE(2019, 7, 31), 1)                                            | Type of the returned value: `CellValueDetailedType.NUMBER_DATE` (non-compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard)            | Cell auto-formatted as **date**                                                                                        | Cell auto-formatted as **regular number**                |
```

**Review by warpech — CHANGES_REQUESTED** [review]
See some ideas about presenting it in a milder and more consistent way. 

Thanks for adding more the tests!

File: docs/guide/built-in-functions.md (hunk 1 above)
Comment (warpech) [reply docs/guide/built-in-functions.md]: The rendering of the duplicate footnote looks weird to me (`[1:1]`), but I guess this is some convention that I was not aware of. It is certainly better.

File: docs/guide/list-of-differences.md [hunk 6]
```diff
@@ -17,6 +17,10 @@ See a full list of differences between HyperFormula, Microsoft Excel, and Google
...
+| EDATE function                                     | =EDATE(DATE(2019, 7, 31), 1)                                              | Type of the returned value: `CellValueDetailedType.NUMBER_DATE` (non-compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard)            | Cell auto-formatted as **date**                                                                                        | Cell auto-formatted as **regular number**                |
+| EOMONTH function                                   | =EOMONTH(DATE(2019, 7, 31), 1)                                            | Type of the returned value: `CellValueDetailedType.NUMBER_DATE` (non-compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard)            | Cell auto-formatted as **date**                                                                                        | Cell auto-formatted as **regular number**                |
```
Comment (warpech) [inline docs/guide/list-of-differences.md:23]: Here, we could soften a bit (same for EDATE):

```suggestion
| EOMONTH function                                   | =EOMONTH(DATE(2019, 7, 31), 1)                                            | Type of the returned value: `CellValueDetailedType.NUMBER_DATE`. This is non-compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard, which defines the return type as a Number, while describing it as a Date serial number through the function summary.            | Cell auto-formatted as **date**                                                                                        | Cell auto-formatted as **regular number**                |
```

**Review by warpech — CHANGES_REQUESTED** [review]
Thank you for the edits! Some more nitpicking:

**Comment by warpech** [comment]
Maybe we could contact the specification editors about this discrepancy by email, or the mailing list (there were [some questions](https://markmail.org/search/?q=EOMONTH&q=list%3Aorg.oasis-open.lists.office-comment) already), or the [issue board](https://issues.oasis-open.org/browse/OFFICE-2027?jql=text%20~%20%22eomonth%22).

File: docs/guide/built-in-functions.md [hunk 7]
```diff
@@ -68,34 +68,34 @@ Total number of functions: **{{ $page.functionsCount }}**
...
+| DAYS             | Calculates the difference between two date values.                                                                                                                                                                                                                          | DAYS(Date2, Date1)                                  |
+| DAYS360          | Calculates the difference between two date values in days, in 360-day basis.                                                                                                                                                                                                | DAYS360(Date2, Date1[, Format])                     |
+| EDATE            | Shifts the given startdate by given number of months and returns it as the number of full days since [nullDate](../api/interfaces/configparams.md#nulldate).<br>Return type: `CellValueDetailedType.NUMBER_DATE` [^non-odff]                                                | EDATE(Startdate, Months)                            |
```
[the remark this answers is not in the document]
Comment (sequba) [reply docs/guide/built-in-functions.md]: Changed

File: docs/guide/built-in-functions.md [hunk 8]
```diff
@@ -507,3 +507,5 @@ Total number of functions: **{{ $page.functionsCount }}**
 | UNICHAR     | Returns the character created by using provided code point.                                                                                                                                                                                       | UNICHAR(Number)                                    |
 | UNICODE     | Returns the Unicode code point of a first character of a text.                                                                                                                                                                                    | UNICODE(Text)                                      |
 | UPPER       | Returns text converted to uppercase.                                                                                                                                                                                                              | UPPER(Text)                                        |
+
+[^non-odff]: Non-compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard. See [compatibility notes](list-of-differences.md).
```
[the remark this answers is not in the document]
Comment (sequba) [reply docs/guide/built-in-functions.md]: I changed it to
```
[^non-odff]: The return value of this function is compliant with the [OpenDocument standard](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html), but the return type is not. See [compatibility notes](list-of-differences.md).
```

**Comment by sequba** [comment]
@warpech 
> Maybe we could contact the specification editors about this discrepancy

I posted a question to their mailing list [office-comment@lists.oasis-open.org](mailto:office-comment@lists.oasis-open.org):

> I'm a developer working on HyperFormula - a javascript formula calculation engine: https://hyperformula.handsontable.com/
> 
> We try to follow the OpenFormula 1.3 specification: https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html
> 
> I have a question about functions EDATE and EOMONTH. The specification states that their return type should be Number, but for me, it seems more logical to return Date from these 2 functions. The numbers returned by EDATE and EOMONTH are not "just numbers" they represent dates in the same way as in functions TODAY, DATE, etc.
> 
> Link to the discussion on that topic in our project repository: https://github.com/handsontable/hyperformula/issues/1083

File: docs/guide/built-in-functions.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (warpech) [reply docs/guide/built-in-functions.md]: Great, thanks!

**Review by warpech — APPROVED** [review]
Great! Before merging, please consider a page layout improvement: https://github.com/handsontable/hyperformula/pull/1120

**Pull request merged**